### PR TITLE
Fix a bug related to abstract imports and match expressions

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Fixes https://github.com/dafny-lang/dafny/issues/5808

### Description
- Add a call to `new MatchFlattener(reporter).PostResolve(module);` after cloning elements for abstract imports, so `.Flattened` is set.
- Fix the behavior of the `afterChildren` parameter in `Node.Visit`, so that closure is actually only called after the children are visited.

### How has this been tested?
Added a CLI test

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Resolver/ModuleResolver.cs [hunk 1]
```diff
@@ -811,38 +811,55 @@ public void RegisterByMethod(Function f, TopLevelDeclWithMembers cl) {
...
+       
+        Cloning with resolved fields is not an option, 
+        because then internal references of the cloned code can point to the old code.
+       
+        I(keyboardDrummer) think it would be better altogether if no cloning was done for abstract modules,
+        But until that happens here is code that explicitly calls MatchFlattener which sets .Flattened
+        Alternatively, we could call all the rewriter.PostResolve methods
+      */
+
+      new MatchFlattener(reporter).PostResolve(module);
```
Comment (keyboardDrummer) [inline Source/DafnyCore/Resolver/ModuleResolver.cs:857]: This added line is the only semantic change in this PR

File: Source/DafnyCore/Generic/Node.cs [hunk 2]
```diff
@@ -144,37 +145,44 @@ public virtual IEnumerable<Assumption> Assumptions(Declaration decl) {
...
           continue;
         }
 
-        if (nodeAfterChildren == null) {
-          toVisit.AddLast(child);
-        } else {
-          toVisit.AddBefore(nodeAfterChildren, child);
+        if (afterChildren != null) {
+          void AfterNodeChildren() => afterChildren(currentNode);
+          toVisit.AddFirst((Action)AfterNodeChildren);
```
Comment (robin-aws) [inline Source/DafnyCore/Generic/Node.cs:166]: Nice fix, funny how simple and clean the original looks while also being completely wrong. :)

File: Source/DafnyCore/Resolver/ModuleResolver.cs [hunk 3]
```diff
@@ -811,38 +811,55 @@ public void RegisterByMethod(Function f, TopLevelDeclWithMembers cl) {
...
+      /* A bug we ran into was that cloning done for abstract modules,
+        did not clone the resolved field,
+        so the .Flattened field of NestedMatchExpr was not set.
+        Also, rewriters.PostResolve was not run, so .Flattened was not set after cloning
+        which led to a crash during Boogie generation.
+       
+        Cloning with resolved fields is not an option, 
+        because then internal references of the cloned code can point to the old code.
+       
+        I(keyboardDrummer) think it would be better altogether if no cloning was done for abstract modules,
```
Comment (robin-aws) [inline Source/DafnyCore/Resolver/ModuleResolver.cs:852]: Out of curiosity what alternative do you imagine? Given the semantics of abstract modules are so close to the idea of making copies of the code, could the better path be doubling-down on clone testing? I don't feel like cloning is going away any time soon.

**Review by robin-aws — APPROVED** [review]

File: Source/DafnyCore/Resolver/ModuleResolver.cs (hunk 3 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/Resolver/ModuleResolver.cs:852]: Whatever operation you want to do on both the abstract module and the original module, pass in a context object that allows you to differentiate between the two scenarios.

**Pull request merged**